### PR TITLE
Allow Initializable versions greater than 256

### LINKED ISSUE #3924: Allow reinitializer versions greater than 256
To maintain storage compatibility with the original Initializable, we made reinitializer versions of type `uint8`. We could use a larger type here once we break compatibility. It still makes sense to pack the value with the `initializing` boolean, so it can't be more than uint248, but even that sounds like an exaggeration and we might want to make it a smaller value like uint64 for the possibility of adding some other values in the same storage slot in the future.

Suggested in https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3870.

### PR DESCRIPTION
**Pull request by ernestognw**
Fixes LIB-828
Fixes #3924 

A greater initialized was proposed for #3924, which is implemented within this PR. Also, the EIP-7201 for Namespaced storage is used here to reduce the risk of storage collisions in upgradeable contracts that are also initializable. In this way, the Initializable variables are placed at a storage location that's guaranteed not to clash.

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [x] Changeset entry (run `npx changeset add`)

### CODE REVIEW

**Comment by ernestognw** [comment]
Bumping pragma to 0.8.20 on Initializable to support NatSpec in structs.

Also, keeping it as a draft at the moment because I suspect we may require a test for storage collisions on storage slot 0.

File: contracts/proxy/utils/Initializable.sol
Comment (ernestognw) [inline contracts/proxy/utils/Initializable.sol:105]: There are several of these for the `$` storage variable. We either keep this rule or wait on https://github.com/protofire/solhint/pull/453 to be merged if that's the case.

**Comment by frangio** [comment]
I like this but before reviewing I'll let @Amxx comment whether he thinks this is a good idea.

File: contracts/proxy/utils/Initializable.sol
Comment (Amxx) [inline contracts/proxy/utils/Initializable.sol:158]: That is not changed by this PR, but I'm wondering if that is the right error to trigger.

File: contracts/proxy/utils/Initializable.sol
Comment (Amxx) [inline contracts/proxy/utils/Initializable.sol:183]: Is `_getInitializableStorage` not a correct name already?

Comment (ernestognw) [reply contracts/proxy/utils/Initializable.sol:183]: It's because of the `$` 😢

File: contracts/proxy/utils/Initializable.sol
Comment (Amxx) [inline contracts/proxy/utils/Initializable.sol:191]: I don't understand why this is necessary

Comment (ernestognw) [reply contracts/proxy/utils/Initializable.sol:191]: Yes, I opened [a PR](https://github.com/protofire/solhint/pull/453) but requires a bit more discussion because they may allow `$` in some other rules.

Comment (frangio) [reply contracts/proxy/utils/Initializable.sol:199]: We should simply get rid of these casing rules and write our own.

**Comment by frangio** [comment]
We discussed offline that we should not include https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3950 in this PR because it breaks a pattern that we use for testing contracts with initializers, as described in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3344.

**Review by frangio — CHANGES_REQUESTED** [review]
We discussed offline that we should not include https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3950 in this PR because it breaks a pattern that we use for testing contracts with initializers, as described in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3344.

**Comment by ernestognw** [comment]
The storage layout checks are failing as expected

File: contracts/proxy/utils/Initializable.sol
Comment (Amxx) [inline contracts/proxy/utils/Initializable.sol:4]: Why do we go to 0.8.20 here ?

File: contracts/proxy/utils/Initializable.sol
Comment (Amxx) [inline contracts/proxy/utils/Initializable.sol:78]: We may want to do the computation differently.

File: test/proxy/utils/Initializable.test.js
Comment (Amxx) [inline test/proxy/utils/Initializable.test.js:216]: Would the test support an hex syntax ? IMO `0xFFFFFFFFFFFFFFFF` would be more readable

File: contracts/proxy/utils/Initializable.sol
Comment (frangio) [reply contracts/proxy/utils/Initializable.sol:4]: `@custom` in structs is only available in 0.8.20. We need to do this in the whole library.

File: contracts/proxy/utils/Initializable.sol
Comment (ernestognw) [reply contracts/proxy/utils/Initializable.sol:78]: Ah, amazing catch! Note we haven't updated EIP-7201 to reflect the impact of Verkle Trees neither.

Are we ok with?:

```suggestion
    // keccak256(abi.encode(uint256(keccak256("openzeppelin.storage.Initializable")) - 256)) & ~bytes32(uint256(0xFF))
    bytes32 private constant _INITIALIZABLE_STORAGE =
        0x2cb0f341752e459a3615b447e07b189024fd0be4129b407c58ddd277d84b6800;
```

File: test/proxy/utils/Initializable.test.js
Comment (ernestognw) [reply test/proxy/utils/Initializable.test.js:216]: Unfortunately not, but we can do `BigInt('0xFFFFFFFFFFFFFFFF').toString()` which is similar imo. What do you think?

File: contracts/proxy/utils/Initializable.sol
Comment (frangio) [reply contracts/proxy/utils/Initializable.sol:78]: Let's not change this until we change the EIP.

File: test/proxy/utils/Initializable.test.js
Comment (frangio) [reply test/proxy/utils/Initializable.test.js]: `(2n ** 64n - 1n).toString()` might be slightly better.

Comment (ernestognw) [reply test/proxy/utils/Initializable.test.js]: I like this one better. Updated.

**Comment by frangio** [comment]
Closed accidentally.

**Comment by ernestognw** [comment]
> Closed accidentally.

`reinitialize(2)`

File: test/proxy/utils/Initializable.test.js
Comment (Amxx) [inline test/proxy/utils/Initializable.test.js:216]: we should consider generalizing this syntax.

**Comment by Amxx** [comment]
LGTM. Do we wait on the linter rule or do we merge like this and come back to it later?

File: contracts/proxy/utils/Initializable.sol
Comment (ernestognw) [reply contracts/proxy/utils/Initializable.sol:186]: I agree with you but we may decide on this once we decide to tackle #3950 as well

File: test/proxy/utils/Initializable.test.js
Comment (ernestognw) [reply test/proxy/utils/Initializable.test.js]: I used `web3.utils.toBN(1).shln(64).subn(1).toString()` instead because I think we've previously agreed on a similar one in the ERC2771Forwarder (but for `uint48`), does this work?

**Comment by ernestognw** [comment]
> LGTM. Do we wait on the linter rule or do we merge like this and come back to it later?

I'd rather not wait on them, I'm sure we'll revisit after starting with the transpiler work

File: .changeset/thick-pumpkins-exercise.md
Comment (frangio) [inline .changeset/thick-pumpkins-exercise.md:5]:
```suggestion
`Initializable`: Use the namespaced storage pattern to avoid putting critical variables in slot 0. Allow reinitializer versions greater than 256.
```

**Review by frangio — APPROVED** [review]

**Pull request merged**
